File: Algoritmos De Ordenamiento/Algoritmos de Otdenamiento in C++/Selection Sort.cpp
#include <string>
#include <chrono>
#include <fstream>
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
#include <functional>
#include <sstream>
#include <iomanip> // Para std::setprecision
#include <dirent.h> // Para manejar directorios

// Algoritmo de ordenamiento: Selection Sort
void selectionSort(std::vector<int>& arr) {
    for (size_t i = 0; i < arr.size() - 1; i++) {
        size_t min_idx = i;
        for (size_t j = i + 1; j < arr.size(); j++) {
            if (arr[j] < arr[min_idx]) {
                min_idx = j;
            }
        }
        std::swap(arr[min_idx], arr[i]);
    }
}

// Leer números de archivo
std::vector<int> readNumbersFromFile(const std::string& file_path) {
    std::ifstream file(file_path);
    std::vector<int> numbers;
    std::string line;

    if (!file) {
        std::cerr << "Error al abrir el archivo: " << file_path << std::endl;
        return numbers; // Retorna un vector vacío si no se puede abrir el archivo
    }

    while (std::getline(file, line)) {
        // Remover corchetes y espacios
        line.erase(remove(line.begin(), line.end(), '['), line.end());
        line.erase(remove(line.begin(), line.end(), ']'), line.end());
        line.erase(remove(line.begin(), line.end(), ' '), line.end());

        std::stringstream ss(line);
        int num;
        char comma; // Variable para leer las comas

        // Leer números separados por comas
        while (ss >> num) {
            numbers.push_back(num);
            ss >> comma; // Ignorar la coma
        }
    }

    std::cout << "Número de elementos en el archivo " << file_path << ": " << numbers.size() << std::endl;

    return numbers;
}

// Medir el tiempo de ejecución del algoritmo de ordenamiento
double measureSortTime(const std::function<void(std::vector<int>&)>& sort_func, std::vector<int> numbers) {
    auto start = std::chrono::high_resolution_clock::now();
    sort_func(numbers);
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> duration = end - start;
    return duration.count();
}

// Guardar resultados en CSV
void saveResultsToCSV(const std::map<std::string, double>& results, const std::string& output_file) {
    std::ofstream csvFile(output_file);
    csvFile << "Archivo,Selection Sort\n";

    for (const auto &result : results) {
        csvFile << result.first << "," << result.second << "\n";
    }

    csvFile.close();
}

// Función principal
int main() {
    std::string folder_path = "C:/Users/saulj/Downloads/ADA_DATOS-main";  // Ruta de la carpeta con los archivos txt

    std::map<std::string, double> results;

    // Obtener archivos .txt en el directorio
    DIR* dir;
    struct dirent* ent;
    if ((dir = opendir(folder_path.c_str())) != nullptr) {
        while ((ent = readdir(dir)) != nullptr) {
            std::string file_name = ent->d_name;
            if (file_name.find(".txt") != std::string::npos) {
                std::string file_path = folder_path + "\\" + file_name; // Agregar la ruta completa
                std::vector<int> numbers = readNumbersFromFile(file_path);

                double time_taken = measureSortTime(selectionSort, numbers);
                results[file_name] = time_taken;
            }
        }
        closedir(dir);
    } else {
        std::cerr << "No se pudo abrir el directorio\n";
    }

    // Guardar resultados en CSV
    saveResultsToCSV(results, "resultados_selection_sort.csv");

    std::cout << "Resultados guardados en resultados_selection_sort.csv" << std::endl;

    // Imprimir resultados
    for (const auto& result : results) {
        std::cout << result.first << " : " << std::fixed << std::setprecision(6) << result.second << " segundos\n";
    }

    return 0;
}

File: Algoritmos De Ordenamiento/Algoritmos de ordenamiento Python/Selection sort.py
import time
import os
import csv

# Algoritmo de ordenamiento: Selection Sort
def selection_sort(arr):
    for i in range(len(arr)):
        min_idx = i
        for j in range(i + 1, len(arr)):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr

def read_numbers_from_file(file_path):
    with open(file_path, 'r') as f:
        content = f.read()
        cleaned_content = ''.join(c for c in content if c.isdigit() or c.isspace() or c == '-')
        numbers = list(map(int, cleaned_content.split()))
    return numbers

def measure_sort_time(sort_function, arr):
    start_time = time.time()
    sort_function(arr.copy())
    end_time = time.time()
    return end_time - start_time

def save_results_to_csv(results, output_file):
    with open(output_file, 'w', newline='') as csvfile:
        fieldnames = ['Archivo', 'Selection Sort']
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

        writer.writeheader()
        for file, timings in results.items():
            row = {'Archivo': file, 'Selection Sort': timings['Selection Sort']}
            writer.writerow(row)

def main():
    folder_path = r'C:\Users\saulj\Downloads\ADA_DATOS-main'
    
    results = {}

    # Iterar sobre los archivos
    for filename in os.listdir(folder_path):
        if filename.endswith('.txt'):
            file_path = os.path.join(folder_path, filename)
            numbers = read_numbers_from_file(file_path)
            duration = measure_sort_time(selection_sort, numbers)
            results[filename] = {'Selection Sort': duration}

    save_results_to_csv(results, 'resultados_python.csv')

    for file, timings in results.items():
        print(f"Resultados para {file}:")
        print(f"Selection Sort: {timings['Selection Sort']:.6f} segundos")
        print()

if __name__ == "__main__":
    main()
